### PR TITLE
[RISCV] Introduce the RISCVLoopIdiomRecognizePass

### PR DESCRIPTION
**Pull request by mshockwave**
RISCVLoopIdiomRecognize pattern matches a specific kind of byte compare loop that looks like:
```
while (i != max_len)
  if (a[i] != b[i])
      break;

... use index i ...
```
And effectively vectorizes it. This is similar to AArch64's AArch64LoopIdiomTransform Pass, except we're using VP intrinsic here.

This improves 557.xz_r in SPEC2017 by nearly 20% (on refrate workload) in terms of dynamic instruction counts.

Note that at this moment, we're not enabling this by default.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp [hunk 1]
```diff
@@ -0,0 +1,752 @@
...
+    "disable-riscv-loop-idiom-bytecmp", cl::Hidden, cl::init(false),
+    cl::desc("Proceed with RISCV Loop Idiom Recognize Pass, but do "
+             "not convert byte-compare loop(s)."));
+
+// CustomLoopIdiomLMUL can be used to customize LMUL for vectorizing loops.
+// It uses the exponent value to represent LMUL i.e. 0 -> LMUL 1, 1 -> LMUL 2, 2
+// -> LMUL 4, 3 -> LMUL 8, etc.
+static cl::opt<unsigned>
+    CustomLoopIdiomLMUL("riscv-loop-idiom-lmul", cl::Hidden, cl::init(1),
+                        cl::Optional,
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp:43]: Is  cl::Optional the default behavior?

File: llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp [hunk 2]
```diff
@@ -0,0 +1,752 @@
...
+      : SubPattern(SP), L(L) {}
+
+  template <typename ITy> bool match(ITy *V) {
+    return L->isLoopInvariant(V) && SubPattern.match(V);
+  }
+};
+
+/// Matches if the value is loop-invariant.
+template <typename Ty>
+inline match_LoopInvariant<Ty> m_LoopInvariant(const Ty &M, const Loop *L) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp:165]: Is this used?

File: llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp [hunk 3]
```diff
@@ -0,0 +1,752 @@
...
+    if (!Index || !Index->getType()->isIntegerTy(32) ||
+        !match(Index, m_c_Add(m_Specific(PN), m_One())))
+      return false;
+  }
+
+  for (BasicBlock *BB : LoopBlocks)
+    for (Instruction &I : *BB)
+      if (&I != PN && &I != Index)
+        for (User *U : I.users()) {
+          auto UI = dyn_cast<Instruction>(U);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp:242]: `auto *UI`

File: llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp (hunk 1 above)
Comment (mshockwave) [reply llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp]: That's right. The cl::Optional is removed now.

File: llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp (hunk 2 above)
Comment (mshockwave) [reply llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp]: No, it's removed now.

File: llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp (hunk 3 above)
Comment (mshockwave) [reply llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp]: Fixed.

File: llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp [hunk 4]
```diff
@@ -0,0 +1,732 @@
...
+            PN.addIncoming(PN.getIncomingValueForBlock(BB), CmpBB);
+            break;
+          }
+      }
+    }
+  };
+
+  // Ensure all Phis in the successors of CmpBB have an incoming value from it.
+  FixSuccessorPhis(EndBB);
+  FixSuccessorPhis(FoundBB);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp:723]: AArch64 has an if before the second `FixSuccessorPhis`

File: llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp [hunk 5]
```diff
@@ -0,0 +1,732 @@
...
+
+  // The new CmpBB block isn't part of the loop, but will need to be added to
+  // the outer loop if there is one.
+  if (!CurLoop->isOutermost())
+    CurLoop->getParentLoop()->addBasicBlockToLoop(CmpBB, LI);
+
+  // Update the dominator tree with the new block.
+  DT.addNewBlock(CmpBB, Preheader);
+}
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp:732]: AArch64 verifies isRecursivelyLCSSAForm here

File: llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp [hunk 6]
```diff
@@ -0,0 +1,732 @@
...
+
+  LLVM_DEBUG(dbgs() << DEBUG_TYPE " Scanning: F["
+                    << CurLoop->getHeader()->getParent()->getName()
+                    << "] Loop %" << CurLoop->getHeader()->getName() << "\n");
+
+  return recognizeAndTransformByteCompare();
+}
+
+bool RISCVLoopIdiomRecognize::recognizeAndTransformByteCompare() {
+  if (DisableByteCmp)
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp:150]: Should we early out if there is no V instructions?

**Comment by ppenzin** [comment]
> As a starting point, have you considered simply generalizing the aarch64 pass into a shared pass which can run on both targets? This would get you code generation via masked.loads for this case - but not the VP & VL predicated form.

I think it makes sense to have a pass separate from AArch64, the semantics of SVE and RVV are different enough that these passes are bound to diverge at some point.

**Comment by topperc** [comment]
> > As a starting point, have you considered simply generalizing the aarch64 pass into a shared pass which can run on both targets? This would get you code generation via masked.loads for this case - but not the VP & VL predicated form.
> 
> I think it makes sense to have a pass separate from AArch64, the semantics of SVE and RVV are different enough that these passes are bound to diverge at some point.

All of the idiom detection code is currently identical. There are some minor differences that I think came from the timing of when we copied the code and later changes made by AArch64. It was not intended for them to diverge. This is one problem having a single pass solves.

There are differences in how we expand the code. Maybe we can use the TTI hooks the middle loop vectorizer uses to have different strategies for AArch64 and RISC-V. If that gets too complicated we could have a common base class that AArch64 and RISC-V both inherit their own passes from so they can share the code that makes sense.

**Comment by ppenzin** [comment]
> There are differences in how we expand the code. Maybe we can use the TTI hooks the middle loop vectorizer uses to have different strategies for AArch64 and RISC-V. If that gets too complicated we could have a common base class that AArch64 and RISC-V both inherit their own passes from so they can share the code that makes sense.

Yeah, that makes sense.

File: llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp [hunk 7]
```diff
@@ -0,0 +1,732 @@
...
+  auto LoopBlocks = CurLoop->getBlocks();
+  // The first block in the loop should contain only 4 instructions, e.g.
+  //
+  //  while.cond:
+  //   %res.phi = phi i32 [ %start, %ph ], [ %inc, %while.body ]
+  //   %inc = add i32 %res.phi, 1
+  //   %cmp.not = icmp eq i32 %inc, %n
+  //   br i1 %cmp.not, label %while.end, label %while.body
+  //
+  auto CondBBInsts = LoopBlocks[0]->instructionsWithoutDebug();
```
[the remark this answers is not in the document]
Comment (mshockwave) [reply llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp:184]: I have addressed this in the new PR(s).

File: llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp [hunk 8]
```diff
@@ -0,0 +1,732 @@
...
+  // while.body:
+  //   %idx = zext i32 %inc to i64
+  //   %idx.a = getelementptr inbounds i8, ptr %a, i64 %idx
+  //   %load.a = load i8, ptr %idx.a
+  //   %idx.b = getelementptr inbounds i8, ptr %b, i64 %idx
+  //   %load.b = load i8, ptr %idx.b
+  //   %cmp.not.ld = icmp eq i8 %load.a, %load.b
+  //   br i1 %cmp.not.ld, label %while.cond, label %while.end
+  //
+  auto LoopBBInsts = LoopBlocks[1]->instructionsWithoutDebug();
```
[the remark this answers is not in the document]
Comment (mshockwave) [reply llvm/lib/Target/RISCV/RISCVLoopIdiomRecognize.cpp:199]: Fixed in the new PR(s).

**Comment by mshockwave** [comment]
Closed in favor of #94081 and #94082

**Pull request closed without merging**